File: backend/src/kb_backend/routers/dimension.py
from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from ..db import get_db
from ..envelope import BusinessError, envelope
from ..models.answer import Answer
from ..models.dimension import DimensionDefinition
from ..schemas.dimension import DimensionAdminOut, DimensionCreate, DimensionOut, DimensionUpdate

router = APIRouter(tags=["dimension"])

_DUPLICATE_KEY_MSG = "维度已存在，请使用其他名称"
_NOT_FOUND_MSG = "维度不存在"
_MYSQL_ER_DUP_ENTRY = 1062


def _raise_if_duplicate_key(exc: IntegrityError) -> None:
    # Mirrors knowledge_base.py's _raise_if_duplicate_name — only translate
    # an actual "duplicate entry" violation; any other integrity error
    # re-raises as-is.
    orig_args = getattr(exc.orig, "args", ())
    if orig_args and orig_args[0] == _MYSQL_ER_DUP_ENTRY:
        raise BusinessError(_DUPLICATE_KEY_MSG, status_code=400) from exc
    raise exc


def _get_or_404(db: Session, key: str) -> DimensionDefinition:
    dim = db.get(DimensionDefinition, key)
    if dim is None:
        raise BusinessError(_NOT_FOUND_MSG, status_code=404)
    return dim


def _answer_count_for_key(db: Session, key: str) -> int:
    # One-key slice of the same Python-side counting approach as
    # list_dimensions_admin below (design doc §4.4) — avoids building a
    # JSON path expression from arbitrary admin-typed text for what would
    # otherwise be a single-dimension count on every create/update/
    # activate/deactivate response.
    coords = db.execute(select(Answer.coord).where(Answer.revoked.is_(False))).scalars().all()
    return sum(1 for coord in coords if key in coord)


def _to_admin_out(dim: DimensionDefinition, answer_count: int) -> DimensionAdminOut:
    return DimensionAdminOut(
        key=dim.key,
        label=dim.label,
        field_type=dim.field_type,
        weight=dim.weight,
        default_value=dim.default_value,
        status=dim.status,
        answer_count=answer_count,
    )


def _set_status(db: Session, key: str, status: str) -> dict:
    dim = _get_or_404(db, key)
    # Mirrors knowledge_base.py's own _set_status: skip the write entirely
    # when already at the target status, instead of committing a no-op on
    # every idempotent activate/deactivate call. Kimi 终审 finding on PR #25.
    if dim.status != status:
        dim.status = status
        db.commit()
        db.refresh(dim)
    # dim.key (the canonical stored spelling), not the raw `key` path
    # param — dimension_definition.key's collation is case/accent
    # insensitive (same as knowledge_base.name), so _get_or_404 can
    # resolve e.g. "region" to a row actually stored as "Region". Answers'
    # coord always uses the canonical spelling; counting against the raw
    # URL spelling would silently report 0 for any request that used a
    # collation-equivalent but differently-spelled key. Codex outer-gate
    # finding on PR #25.
    out = _to_admin_out(dim, _answer_count_for_key(db, dim.key))
    return envelope(out.model_dump(mode="json"))


@router.get("/dimensions")
def list_dimensions(db: Session = Depends(get_db)) -> dict:
    rows = (
        db.execute(
            select(DimensionDefinition)
            .where(DimensionDefinition.status == "active")
            .order_by(DimensionDefinition.key)
        )
        .scalars()
        .all()
    )
    out = [DimensionOut.model_validate(row) for row in rows]
    return envelope([o.model_dump(mode="json") for o in out])


@router.get("/admin/dimensions")
def list_dimensions_admin(db: Session = Depends(get_db)) -> dict:
    """Internal management view (design doc §2/§4.6) — every dimension
    regardless of status, plus `default_value`/`status`/`answer_count`,
    unlike the external, active-only /dimensions above. Deliberately a
    separate path (not /dimensions/admin) so it can never collide with a
    future /dimensions/{key} lookup, given `key` is arbitrary admin-typed
    text with no reserved words (design doc §4.1/§4.6)."""
    rows = db.execute(select(DimensionDefinition).order_by(DimensionDefinition.key)).scalars().all()

    # One pass over every non-revoked answer's coord, counted in Python
    # rather than one JSON_CONTAINS_PATH query per dimension — `key` is
    # arbitrary admin-typed text (could contain '.', '"', '$', etc., all
    # meaningful in JSON path syntax), so building a path expression by
    # string interpolation is fragile at best. Design doc §4.4.
    answer_counts: dict[str, int] = {}
    coords = db.execute(select(Answer.coord).where(Answer.revoked.is_(False))).scalars().all()
    for coord in coords:
        for key in coord:
            answer_counts[key] = answer_counts.get(key, 0) + 1

    out = [_to_admin_out(row, answer_counts.get(row.key, 0)) for row in rows]
    return envelope([o.model_dump(mode="json") for o in out])


@router.post("/dimensions")
def create_dimension(payload: DimensionCreate, db: Session = Depends(get_db)) -> dict:
    # design doc §4.1: whatever the admin types as `label` becomes `key`
    # verbatim, unchanged, forever — no slugify/transliteration.
    dim = DimensionDefinition(
        key=payload.label,
        label=payload.label,
        field_type=payload.field_type,
        weight=payload.weight,
        default_value=payload.default_value,
        status="active",
    )
    db.add(dim)
    try:
        db.commit()
    except IntegrityError as exc:
        db.rollback()
        _raise_if_duplicate_key(exc)
    db.refresh(dim)

    # A brand-new dimension has never been referenced by any answer —
    # genuinely 0, not a placeholder, so no count query is needed here.
    out = _to_admin_out(dim, answer_count=0)
    return envelope(out.model_dump(mode="json"))


@router.patch("/dimensions/{key}")
def update_dimension(key: str, payload: DimensionUpdate, db: Session = Depends(get_db)) -> dict:
    dim = _get_or_404(db, key)

    if payload.label is not None:
        dim.label = payload.label
    if payload.weight is not None:
        dim.weight = payload.weight
    # "default_value omitted" (keep unchanged) and "default_value explicitly
    # null" (clear it) are both meaningful and distinct — check
    # model_fields_set, not the value, mirroring update_knowledge_base's
    # own handling of `description`.
    if "default_value" in payload.model_fields_set:
        dim.default_value = payload.default_value

    db.commit()
    db.refresh(dim)

    out = _to_admin_out(dim, _answer_count_for_key(db, dim.key))
    return envelope(out.model_dump(mode="json"))


@router.post("/dimensions/{key}/activate")
def activate_dimension(key: str, db: Session = Depends(get_db)) -> dict:
    return _set_status(db, key, "active")


@router.post("/dimensions/{key}/deactivate")
def deactivate_dimension(key: str, db: Session = Depends(get_db)) -> dict:
    return _set_status(db, key, "deprecated")

File: frontend/src/api/dimension.ts
/**
 * 维度管理（字段镜像 backend/src/kb_backend/schemas/dimension.py）。
 * 维度定义全局共享；每个知识库单独「启用」后才能在该库的答案条件里使用。
 */
import { request } from '@/utils/request'

/** 对外（启用中）维度：DimensionOut */
export interface Dimension {
  key: string
  label: string
  field_type: 'text' | 'number' | 'date' | 'boolean'
  weight: number
}

/** 管理侧维度：DimensionAdminOut——不分状态，含对外视图刻意省略的字段 */
export interface AdminDimension extends Dimension {
  default_value: string | null
  status: 'active' | 'deprecated'
  answer_count: number
}

/**
 * 与 DimensionUpdateInput 结构上分开，镜像后端 DimensionCreate/DimensionUpdate
 * 的拆分——DimensionUpdate 根本没有 field_type 字段（设计文档 §3.2，issue #13），
 * 不是「送了会被拒」而是不存在。
 */
export interface DimensionCreateInput {
  label: string
  field_type: Dimension['field_type']
  weight: number
  default_value: string | null
}

export interface DimensionUpdateInput {
  label: string
  weight: number
  /**
   * 永远显式发送、从不省略——null 表示「清空」，非空字符串表示「设为该值」。
   * 整个省略这个 key 会被后端基于 model_fields_set 的逻辑理解为「保持不变」，
   * 用户刚清空的场景会静默不生效（设计文档 §4.2，issue #13）。
   */
  default_value: string | null
}

/** 某知识库已启用的维度（6.x /knowledge-bases/{id}/enabled-dimensions） */
export function listEnabledDimensions(kbId: number) {
  return request.get<Dimension[]>(`/knowledge-bases/${kbId}/enabled-dimensions`)
}

/**
 * 某维度在该知识库现存答案条件里出现过的全部取值（条件选择器的下拉候选，
 * 仅对 text 维度调用）。dimension_key 走 axios params 自动编码——维度 key
 * 是任意文本，裸拼 URL 的坑同 updateDimension 注释。silent：候选只是输入
 * 辅助，拉取失败不弹全局错误提示，由选择器内联降级为纯手输。
 */
export function listDimensionValues(kbId: number, dimensionKey: string) {
  return request.get<string[]>(`/knowledge-bases/${kbId}/dimension-values`, {
    params: { dimension_key: dimensionKey },
    silent: true,
  })
}

/** 管理侧维度全量列表（含已停用） */
export function listAdminDimensions() {
  return request.get<AdminDimension[]>('/admin/dimensions')
}

/** 新增全局维度 */
export function createDimension(input: DimensionCreateInput) {
  return request.post<AdminDimension>('/dimensions', input)
}

/**
 * 编辑全局维度。encodeURIComponent(key)：维度 key 是用户输入的 label 原文
 * （后端只拒绝其中的 "/"，设计文档 §4.3），"?"/"#"/"&" 都合法，裸拼进 URL
 * 会被解析成查询串/片段，静默打到错误资源（Codex 结论，PR #29）。
 */
export function updateDimension(key: string, input: DimensionUpdateInput) {
  return request.patch<AdminDimension>(`/dimensions/${encodeURIComponent(key)}`, input)
}

/** 启用/停用全局维度 */
export function setDimensionStatus(key: string, status: 'active' | 'deprecated') {
  return request.post<AdminDimension>(
    `/dimensions/${encodeURIComponent(key)}/${status === 'active' ? 'activate' : 'deactivate'}`,
  )
}

/** 整体设置某知识库启用的维度集合 */
export function setEnabledDimensions(kbId: number, dimensionKeys: string[]) {
  return request.put<Dimension[]>(`/knowledge-bases/${kbId}/enabled-dimensions`, { dimension_keys: dimensionKeys })
}
